builder/builder.go

```diff
@@ -199,15 +204,143 @@ func (b *Builder) startRegistryProxy(ctx context.Context, caps apicaps.CapSet) (
+// special host name, host.docker.internal, is made available to access the
+// host's network. Docker can only pull insecurely from localhost, so we use a
+// socat container to proxy localhost:<port> request back out to the local
+// registry proxy created above.
+func (b *Builder) startDarwinProxyHelper(ctx context.Context, registryPort int) (int, error) {
+	infoMap, err := b.opt.ContainerFrontend.ContainerInfo(ctx, darwinProxyContainerName)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to check info")
+	}
+
+	var exists bool
+	if info, ok := infoMap[darwinProxyContainerName]; ok && info.Status != "missing" {
+		exists = true
+	}
+
+	if exists {
+		err := b.opt.ContainerFrontend.ContainerRemove(ctx, true, darwinProxyContainerName)
+		if err != nil {
+			return 0, errors.Wrap(err, "failed to remove existing container")
+		}
+	}
+
+	runCfg := containerutil.ContainerRun{
+		NameOrID: darwinProxyContainerName,
+		ImageRef: "alpine/socat:1.7.4.4",
+		Ports: []containerutil.Port{
+			{
+				IP:            "127.0.0.1",
+				HostPort:      0, // Bind to available port
+				ContainerPort: 80,
+				Protocol:      containerutil.ProtocolTCP,
+			},
+		},
+		ContainerArgs: []string{
+			"tcp-listen:80,fork,reuseaddr",
+			fmt.Sprintf("tcp:host.docker.internal:%d", registryPort),
+		},
+	}
+
+	err = b.opt.ContainerFrontend.ContainerRun(ctx, runCfg)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to start support container")
+	}
+
+	infoMap, err = b.opt.ContainerFrontend.ContainerInfo(ctx, darwinProxyContainerName)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to check info")
+	}
+
+	info, ok := infoMap[darwinProxyContainerName]
+	if !ok {
+		return 0, errors.New("container does not exist")
+	}
+
+	if len(info.Ports) == 0 {
+		return 0, errors.New("container port not set")
+	}
+
+	parts := strings.Split(info.Ports[0], ":")
+	if len(parts) != 3 {
+		return 0, errors.Errorf("invalid container port format: %s", info.Ports[0])
+	}
+
+	assignedPort, err := strconv.Atoi(parts[1])
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to parse container port")
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	// Wait for the proxy chain to resolve to the BK registry. The /v2/ path
+	// will return a 200 when ready.
+	for {
+		res, err := http.Get(fmt.Sprintf("http://localhost:%d/v2/", assignedPort))
+		if res != nil && res.Body != nil {
+			res.Body.Close()
+		}
+		if err == nil && res != nil && res.StatusCode == http.StatusOK {
+			break
+		}
+		select {
+		case <-ctx.Done():
+			return 0, ctx.Err()
+		case <-time.After(time.Second):
+			continue
+		}
+	}
+
+	return assignedPort, nil
+}
+
+func (b *Builder) stopDarwinProxyHelper(ctx context.Context) error {
+	err := b.opt.ContainerFrontend.ContainerRemove(ctx, true, darwinProxyContainerName)
+	if err != nil {
+		return errors.Wrap(err, "failed to stop support container")
+	}
+	return nil
```

Comment:
It's probably worth moving all this in its own package. Builder has gotten pretty big and is normally concerned with other things - like putting together the build function and interpreting the build results.

Reply:
Sure, that makes sense.